mpi/base/helpers.hpp

```diff
@@ -0,0 +1,161 @@
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_MPI_HELPERS_HPP_
+#define GKO_MPI_HELPERS_HPP_
+
+
+#include <functional>
+
+
+#include <mpi.h>
+
+
+#include <ginkgo/core/base/exception_helpers.hpp>
+
+
+namespace gko {
+/**
+ * @brief The MPI namespace.
+ *
+ * @ingroup mpi
+ */
+namespace mpi {
+/**
+ * @brief The helpers namespace.
+ *
+ * @ingroup helper
+ */
+namespace helpers {
+
+#define GKO_MPI_DATATYPE(BaseType, MPIType)                                  \
+    inline MPI_Datatype get_mpi_type(const BaseType&) { return MPIType; }    \
+    static_assert(true,                                                      \
+                  "This assert is used to counter the false positive extra " \
+                  "semi-colon warnings")
+
+
+GKO_MPI_DATATYPE(bool, MPI_C_BOOL);
+GKO_MPI_DATATYPE(char, MPI_CHAR);
+GKO_MPI_DATATYPE(unsigned char, MPI_UNSIGNED_CHAR);
+GKO_MPI_DATATYPE(unsigned, MPI_UNSIGNED);
+GKO_MPI_DATATYPE(int, MPI_INT);
+GKO_MPI_DATATYPE(unsigned long, MPI_UNSIGNED_LONG);
+GKO_MPI_DATATYPE(unsigned short, MPI_UNSIGNED_SHORT);
+GKO_MPI_DATATYPE(long, MPI_LONG);
+GKO_MPI_DATATYPE(float, MPI_FLOAT);
+GKO_MPI_DATATYPE(double, MPI_DOUBLE);
+GKO_MPI_DATATYPE(long double, MPI_LONG_DOUBLE);
+GKO_MPI_DATATYPE(std::complex<float>, MPI_COMPLEX);
+GKO_MPI_DATATYPE(std::complex<double>, MPI_DOUBLE_COMPLEX);
+
+
+namespace detail {
+namespace operations {
+
+template <typename ValueType>
+void custom(void* in, void* inout, int* size, MPI_Datatype*)
+{
+    auto l_in = reinterpret_cast<ValueType*>(in);
+    auto l_inout = reinterpret_cast<ValueType*>(inout);
+    ValueType sum = 0.0;
+    for (auto i = 0; i < *size; ++i) {
+    }
+    *l_inout = (*l_in);
+}
+
+
+}  // namespace operations
+}  // namespace detail
+
+
+template <typename ValueType>
+MPI_Op get_operation(gko::mpi::op_type op)
+{
+    switch (op) {
+    case gko::mpi::op_type::sum:
+        return MPI_SUM;
+        break;
+    case gko::mpi::op_type::min:
+        return MPI_MIN;
+        break;
+    case gko::mpi::op_type::max:
+        return MPI_MAX;
+        break;
+    case gko::mpi::op_type::product:
+        return MPI_PROD;
+        break;
+    case gko::mpi::op_type::custom: {
+        // TEMPLATE to create custom operations
+        MPI_Op op;
+        GKO_ASSERT_NO_MPI_ERRORS(
+            MPI_Op_create(detail::operations::custom<ValueType>, true, &op));
+        return op;
+        break;
```

Comment:
I don't see a point in adding this, as 1. the custom operation is a no-op, 2. the custom operation can't be changed by the user.